### ISSUE TITLE
Graphiql > "Explain" is not working with @graphile/pro and --read-only-connection

### ISSUE THREAD

**Issue by mabilbao** [graphile/postgraphile#1408]
### Graphiql > "Explain" is not working for me

Sorry if I am creating this question here and maybe it's not the proper place. I was using Postgraphile since a couple of months ago and something great to check if my queries have performance issues is the option `explain` into `/graphiql`. 

For some reason, my console shows me the following error message each time that I enable that option:
```json
{
  "errors": [
    {
      "message": "pgClient.startExplain is not a function",
      "stack": "TypeError: pgClient.startExplain is not a function\n    at /home/node/project/packages/database/node_modules/postgraphile/src/postgraphile/withPostGraphileContext.ts:217:18\n    at withAuthenticatedPgClient (/home/node/project/packages/database/node_modules/postgraphile/src/postgraphile/withPostGraphileContext.ts:162:24)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)\n    at /home/node/project/packages/database/node_modules/postgraphile/src/postgraphile/http/createPostGraphileHttpRequestHandler.ts:759:24\n    at async Promise.all (index 0)\n    at requestHandler (/home/node/project/packages/database/node_modules/postgraphile/src/postgraphile/http/createPostGraphileHttpRequestHandler.ts:682:17)"
    }
  ]
}
```

I debugged a little bit and it looks like the problem is sometimes the function `startExplain()` (https://github.com/graphile/postgraphile/blob/v4/src/postgraphile/withPostGraphileContext.ts#L214-L239) come into the `pgClient` object and sometimes not, causing an issue not only in the response else in the entire console because I can't get the schema until disabling the explain and reload.

I didn't see any similar issue to this one, do you know something that I could check or something that I could be doing wrong?

Thank you in advance!

#### postgraphilerc
```
const isProduction = process.env.NODE_ENV === 'production'
const options: PostGraphileOptions = {
  appendPlugins: [
    require('@graphile-contrib/pg-many-to-many'),
    require('@graphile-contrib/pg-simplify-inflector'),
    require('postgraphile-plugin-connection-filter'),
    extendSchema
  ],
  dynamicJson: true,
  enableQueryBatching: true,
  graphiql: true,
  enhanceGraphiql: true,
  allowExplain: req => {
    return true
  },

  pgDefaultRole: 'anonymous',
  ignoreRBAC: false,
  setofFunctionsContainNulls: false,
  disableDefaultMutations: false,
  disableQueryLog: isProduction,
  extendedErrors: isProduction
    ? []
    : [
        'severity',
        'code',
        'detail',
        'hint',
        'position',
        'internalPosition',
        'internalQuery',
        'where',
        'schema',
        'table',
        'column',
        'dataType',
        'constraint',
        'file',
        'line',
        'routine',
      ],
  showErrorStack: !isProduction,
  
  jwtVerifyOptions: {
    ignoreExpiration: true
  },
  jwtSecret: process.env.JWT_SECRET,
  jwtPgTypeIdentifier: '"project".jwt_token',
}
```

#### Dependencies
```
  "dependencies": {
    "@graphile-contrib/pg-many-to-many": "^1.0.0-beta.5",
    "@graphile-contrib/pg-simplify-inflector": "^3.0.0",
    "@graphile/pro": "^0.9.0",
    "@sentry/node": "^5.27.2",
    "@sentry/tracing": "^5.27.2",
    "body-parser": "^1.19.0",
    "consola": "^2.10.1",
    "cors": "^2.8.5",
    "dotenv": "^6.2.0",
    "express": "^4.17.1",
    "graphile-build-pg": "^4.10.0",
    "graphile-utils": "^4.10.0",
    "http-errors": "^1.7.3",
    "jsonwebtoken": "^8.5.1",
    "lodash": "^4.17.19",
    "passport": "^0.4.0",
    "passport-jwt": "^4.0.0",
    "passport-saml": "^1.2.0",
    "pg": "^8.5.1",
    "postgraphile": "^4.9.2",
    "postgraphile-plugin-connection-filter": "^2.1.1",
    "ts-node": "9.0.0",
    "typescript": "^3.6.2"
  },
```

Executed in Docker - Node: 14.15.1

**Comment by benjie** [comment]
I was not able to reproduce this with `npx postgraphile -c pggql_test -s a --allow-explain --enhance-graphiql` and don't have time to debug your full setup; can you make a minimal reproduction? Remove everything you can whilst ensuring the issue still occurs.

**Comment by mabilbao** [comment]
Thank you for your answer!

I finally decided to test it without my express server and it's working perfectly well. So trying to find an explanation I decided to remove everything in my node project I finally figured out that my problem was the `readOnlyConnection` that I setting in the Postgraphile connection. So I tried to test it by adding the read replica to my CLI and finally I could get the same error going outside of my express:

https://drive.google.com/file/d/1fNlRB7OFH6pLULHAHdVxYF-6V5hUEjnp/view?usp=sharing
```
npx postgraphile --plugins @graphile/pro \
--read-only-connection 'postgres://local_user:password@localhost:2345/mainschema' \
--connection 'postgres://local_user:password@localhost:2345/mainschema' \
--enhance-graphiql \
--port 5002 \
--subscriptions \
--watch \
--dynamic-json \
--no-setof-functions-contain-nulls \
--no-ignore-rbac \
--show-error-stack=json \
--extended-errors hint,detail,errcode \
--append-plugins @graphile-contrib/pg-simplify-inflector \
--export-schema-graphql schema.graphql \
--graphiql "/graphiql" \
--enhance-graphiql \
--allow-explain \
--enable-query-batching \
--legacy-relations omit \
--schema mainschema \
--default-role 'anonymous' \
--jwt-secret 'secret-123' \
--jwt-verify-ignore-expiration \
--jwt-token-identifier '"mainschema".jwt_token'
```

**Error**
```
{
  "errors": [
    {
      "message": "pgClient.startExplain is not a function",
      "stack": [
        "TypeError: pgClient.startExplain is not a function",
        "    at /Users/martin/Code/projects/packages/database/node_modules/postgraphile/build/postgraphile/withPostGraphileContext.js:162:26",
        "    at withAuthenticatedPgClient (/Users/martin/Code/projects/packages/database/node_modules/postgraphile/build/postgraphile/withPostGraphileContext.js:105:30)",
        "    at processTicksAndRejections (internal/process/task_queues.js:93:5)",
        "    at async /Users/martin/Code/projects/packages/database/node_modules/postgraphile/build/postgraphile/http/createPostGraphileHttpRequestHandler.js:695:34",
        "    at async Promise.all (index 0)",
        "    at async graphqlRouteHandler (/Users/martin/Code/projects/packages/database/node_modules/postgraphile/build/postgraphile/http/createPostGraphileHttpRequestHandler.js:627:23)",
        "    at async /Users/martin/Code/projects/packages/database/node_modules/postgraphile/build/postgraphile/http/createPostGraphileHttpRequestHandler.js:333:17"
      ]
    }
  ],
  "meta": {
    "graphqlQueryCost": 1
  }
}
```

I am not having the issue with this config:

https://drive.google.com/file/d/1erQyaJrqCX-9feeOj8wM2Lr8JXRM_yKq/view?usp=sharing
```
npx postgraphile --plugins @graphile/pro \
--connection 'postgres://local_user:password@localhost:2345/mainschema' \
--enhance-graphiql \
--port 5002 \
--subscriptions \
--watch \
--dynamic-json \
--no-setof-functions-contain-nulls \
--no-ignore-rbac \
--show-error-stack=json \
--extended-errors hint,detail,errcode \
--append-plugins @graphile-contrib/pg-simplify-inflector \
--export-schema-graphql schema.graphql \
--graphiql "/graphiql" \
--enhance-graphiql \
--allow-explain \
--enable-query-batching \
--legacy-relations omit \
--schema mainschema \
--default-role 'anonymous' \
--jwt-secret 'secret-123' \
--jwt-verify-ignore-expiration \
--jwt-token-identifier '"mainschema".jwt_token'
```

**Comment by benjie** [comment]
Reproduced with `postgraphile --plugins @graphile/pro --connection 'postgres:///test' --enhance-graphiql --allow-explain --read-only-connection 'postgres:///test'`

**Comment by benjie** [comment]
Fixed in `@graphile/pro@1.0.0` which I just released :+1:

**Comment by mabilbao** [comment]
Thank you so much @benjie 😄
